lib/block_view_validator.go

```diff
@@ -1680,6 +1704,72 @@ func (bav *UtxoView) IsValidUnjailValidatorMetadata(transactorPublicKey []byte)
 	return nil
 }
 
+func (bav *UtxoView) SanityCheckUnregisterAsValidatorTxn(
+	transactorPKID *PKID,
+	utxoOp *UtxoOperation,
+	amountNanos *uint256.Int,
+) error {
+	if utxoOp.Type != OperationTypeUnregisterAsValidator {
+		return fmt.Errorf("SanityCheckUnregisterAsValidatorTxn: called with %v", utxoOp.Type)
+	}
+
+	// Sanity check the deleted ValidatorEntry.
+	if utxoOp.PrevValidatorEntry == nil {
+		return errors.New("SanityCheckUnregisterAsValidatorTxn: nil PrevValidatorEntry provided")
+	}
+	if !utxoOp.PrevValidatorEntry.ValidatorPKID.Eq(transactorPKID) {
+		return errors.New("SanityCheckUnregisterAsValidatorTxn: ValidatorPKID doesn't match TransactorPKID")
+	}
+	if !utxoOp.PrevValidatorEntry.TotalStakeAmountNanos.Eq(amountNanos) {
+		return errors.New("SanityCheckUnregisterAsValidatorTxn: TotalStakeAmountNanos doesn't match")
+	}
+	currentValidatorEntry, err := bav.GetValidatorByPKID(utxoOp.PrevValidatorEntry.ValidatorPKID)
+	if err != nil {
+		return errors.Wrapf(err, "SanityCheckUnregisterAsValidatorTxn: error retrieving ValidatorEntry: ")
+	}
+	if currentValidatorEntry != nil {
+		return errors.New("SanityCheckUnregisterAsValidatorTxn: ValidatorEntry was not deleted")
+	}
+
+	// Sanity check that there are no existing StakeEntries for the validator.
+	stakeEntries, err := bav.GetStakeEntriesForValidatorPKID(utxoOp.PrevValidatorEntry.ValidatorPKID)
+	if err != nil {
+		return errors.Wrapf(err, "SanityCheckUnregisterAsValidatorTxn: error retrieving StakeEntries: ")
+	}
+	if len(stakeEntries) != 0 {
+		return errors.New("SanityCheckUnregisterAsValidatorTxn: StakeEntries for ValidatorEntry still exist")
+	}
+
+	// Sanity check the deleted StakeEntries.
```

Comment:
There's one more validation I could do here which is to ensure that the created/updated LockedStakeEntries have the correct LockedAmountNanos. I punted on that since we already do a check in `_connectUnregisterAsValidator`, but mostly because we don't have a `GetLockedStakeEntriesForValidatorAndEpoch` getter function. Our LockedStakeEntries getter requires a ValidatorPKID + StakerPKID + EpochNumber, but we would want to retrieve them across StakerPKIDs in this sanity check. I could add such a getter. But it's also non-trivial logic to keep track of the PrevLockedStakeEntry existing LockedAmountNanos versus what was added by this unregistering txn in this sanity check. All do-able but felt like overkill for this sanity check. LMK.